framework/src/libtorch/controls/LibtorchNeuralNetControl.C

```diff
@@ -0,0 +1,250 @@
+  // combinations are not allowed
+  conditionalParameterError("filename",
+                            {"num_neurons_per_layer", "activation_function"},
+                            !getParam<bool>("torch_script_format"));
+
+  if (_response_names.size() != _response_shift_factors.size())
+    paramError("response_shift_factors",
+               "The number of shift factors is not the same as the number of responses!");
+
+  if (_response_names.size() != _response_scaling_factors.size())
+    paramError(
+        "response_scaling_factors",
+        "The number of normalization coefficients is not the same as the number of responses!");
+
+  if (_control_names.size() != _action_scaling_factors.size())
+    paramError("action_scaling_factors",
+               "The number of normalization coefficients is not the same as the number of "
+               "controlled parameters!");
+
+  // We link to the postprocessor values so that we can fetch them any time. This also raises
+  // errors if we don't have the postprocessors requested in the input.
+  for (unsigned int resp_i = 0; resp_i < _response_names.size(); ++resp_i)
+    _response_values.push_back(&getPostprocessorValueByName(_response_names[resp_i]));
+
+  // If the user wants to read the neural net from file, we do it. We can read it from a
+  // torchscript file, or we can create a shell and read back the parameters
+  if (isParamValid("filename"))
+  {
+    std::string filename = getParam<std::string>("filename");
+    if (getParam<bool>("torch_script_format"))
+      _nn = std::make_shared<Moose::LibtorchTorchScriptNeuralNet>(filename);
+    else
+    {
+      unsigned int num_inputs = _response_names.size() * _input_timesteps;
+      unsigned int num_outputs = _control_names.size();
+      std::vector<unsigned int> num_neurons_per_layer =
+          getParam<std::vector<unsigned int>>("num_neurons_per_layer");
+      std::vector<std::string> activation_functions =
+          parameters.isParamSetByUser("activation_function")
+              ? getParam<std::vector<std::string>>("activation_function")
+              : std::vector<std::string>({"relu"});
+      auto nn = std::make_shared<Moose::LibtorchArtificialNeuralNet>(
+          filename, num_inputs, num_outputs, num_neurons_per_layer, activation_functions);
+
+      try
+      {
+        torch::load(nn, filename);
+        _nn = std::make_shared<Moose::LibtorchArtificialNeuralNet>(*nn);
+      }
+      catch (...)
+      {
+        mooseError(
+            "The requested pytorch parameter file could not be loaded. This can either be the"
+            "result of the file not existing or a misalignment in the generated container and"
+            "the data in the file. Make sure the dimensions of the generated neural net are the"
+            "same as the dimensions of the parameters in the input file!");
+      }
+    }
+  }
+}
+
+void
+LibtorchNeuralNetControl::execute()
+{
+  if (_nn)
```

Comment:
When will there not be one?

Reply:
Well, by default Controls are executed at the very beginning of the simulation. In case when the net is transferred to the Controller, there might be occasions in the early execution when the Controller doesn't have a NN yet. This guards against that and expresses that if we don't have a net, we shouldn't tough the system.